File: mobile/test/telephone_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:lgr_parents/outils/telephone.dart';

/// Numéros tchadiens.
///
/// Ces tests doivent rester alignés sur `web/src/lib/telephone.ts` : c'est la
/// même règle, écrite deux fois faute de pouvoir partager du code entre Dart
/// et TypeScript. Un désaccord entre les deux se traduirait par un parent que
/// l'application accepte et que le serveur refuse — ou l'inverse.
void main() {
  group('Validation', () {
    test('accepte les quatre préfixes tchadiens', () {
      for (final prefixe in ['3', '6', '8', '9']) {
        expect(Telephone.valide('${prefixe}1234567'), isTrue, reason: 'préfixe $prefixe');
      }
    });

    test('refuse un préfixe hors plan de numérotation', () {
      for (final prefixe in ['0', '1', '2', '4', '5', '7']) {
        expect(Telephone.valide('${prefixe}1234567'), isFalse, reason: 'préfixe $prefixe');
      }
    });

    test('exige exactement huit chiffres', () {
      expect(Telephone.valide('6612345'), isFalse, reason: 'sept chiffres');
      expect(Telephone.valide('66123456'), isTrue, reason: 'huit chiffres');
      expect(Telephone.valide('661234567'), isFalse, reason: 'neuf chiffres');
    });
  });

  group('Formes acceptées à la saisie', () {
    // Le secrétariat écrit un même numéro de vingt façons. Toutes doivent
    // mener au même numéro, sinon le même parent existe en double au fichier
    // et ne reçoit jamais ses alertes.
    const attendu = '66000000';

    for (final saisie in [
      '66000000',
      '66 00 00 00',
      '66-00-00-00',
      '+23566000000',
      '+235 66 00 00 00',
      '23566000000',
      '0023566000000',
      '066000000',
    ]) {
      test('« $saisie »', () {
        expect(Telephone.partieNationale(saisie), attendu);
        expect(Telephone.valide(saisie), isTrue);
        expect(Telephone.international(saisie), '+235$attendu');
      });
    }
  });

  group('Motif de refus', () {
    test('dit combien de chiffres manquent', () {
      expect(Telephone.motifRefus('660'), contains('5 chiffres'));
      expect(Telephone.motifRefus('6600000'), contains('1 chiffre'));
      expect(Telephone.motifRefus('6600000'), isNot(contains('1 chiffres')));
    });

    test('nomme les préfixes valides', () {
      expect(Telephone.motifRefus('51234567'), contains('3, 6, 8, 9'));
    });

    test('ne dit rien sur un numéro correct', () {
      expect(Telephone.motifRefus('66000000'), isNull);
      expect(Telephone.motifRefus('+235 31 23 45 67'), isNull);
    });

    test('demande une saisie quand le champ est vide', () {
      expect(Telephone.motifRefus(''), isNotNull);
    });
  });

  group('Mise en forme', () {
    test('groupe par deux', () {
      expect(Telephone.lisible('66000000'), '66 00 00 00');
      expect(Telephone.lisible('+23591912191'), '91 91 21 91');
    });

    test('supporte un numéro incomplet', () {
      expect(Telephone.lisible('660'), '66 0');
    });
  });

  group('Formateur de saisie', () {
    final formateur = FormateurTelephone();

    TextEditingValue frapper(String texte) => formateur.formatEditUpdate(
      TextEditingValue.empty,
      TextEditingValue(text: texte),
    );

    test('insère les espaces au fil de la frappe', () {
      expect(frapper('6').text, '6');
      expect(frapper('66').text, '66');
      expect(frapper('660').text, '66 0');
      expect(frapper('66000000').text, '66 00 00 00');
    });

    test('rejette tout ce qui n\'est pas un chiffre', () {
      expect(frapper('66abc00').text, '66 00');
      expect(frapper('+235-66').text, '23 56 6');
    });

    test('tronque au-delà de huit chiffres', () {
      // Un parent qui colle un numéro avec indicatif ne doit pas se retrouver
      // avec douze chiffres dans le champ.
      expect(frapper('660000009999').text, '66 00 00 00');
    });

    test('laisse le curseur en fin de saisie', () {
      final valeur = frapper('66000');
      expect(valeur.selection.baseOffset, valeur.text.length);
    });
  });
}

File: web/src/lib/telephone.test.ts
import assert from "node:assert/strict";
import { describe, it } from "node:test";

import {
  formatInternational,
  formatLisible,
  motifRefus,
  partieNationale,
  telephoneValide,
} from "./telephone.ts";

/**
 * Numéros tchadiens.
 *
 * Ces cas sont les MÊMES que ceux de `mobile/test/telephone_test.dart`. Les
 * deux implémentations doivent rester d'accord : un désaccord se traduirait
 * par un parent que l'application accepte et que le serveur refuse — ou
 * l'inverse, ce qui est pire, car il croirait s'être inscrit.
 */

describe("Validation", () => {
  it("accepte les quatre préfixes tchadiens", () => {
    for (const prefixe of ["3", "6", "8", "9"]) {
      assert.equal(telephoneValide(`${prefixe}1234567`), true, `préfixe ${prefixe}`);
    }
  });

  it("refuse un préfixe hors plan de numérotation", () => {
    for (const prefixe of ["0", "1", "2", "4", "5", "7"]) {
      assert.equal(telephoneValide(`${prefixe}1234567`), false, `préfixe ${prefixe}`);
    }
  });

  it("exige exactement huit chiffres", () => {
    assert.equal(telephoneValide("6612345"), false);
    assert.equal(telephoneValide("66123456"), true);
    assert.equal(telephoneValide("661234567"), false);
  });
});

describe("Formes acceptées à la saisie", () => {
  const attendu = "66000000";

  for (const saisie of [
    "66000000",
    "66 00 00 00",
    "66-00-00-00",
    "+23566000000",
    "+235 66 00 00 00",
    "23566000000",
    "0023566000000",
    "066000000",
  ]) {
    it(`« ${saisie} »`, () => {
      assert.equal(partieNationale(saisie), attendu);
      assert.equal(telephoneValide(saisie), true);
      assert.equal(formatInternational(saisie), `+235${attendu}`);
    });
  }
});

describe("Motif de refus", () => {
  it("dit combien de chiffres manquent", () => {
    assert.match(motifRefus("660") ?? "", /5 chiffres/);
    assert.match(motifRefus("6600000") ?? "", /1 chiffre/);
    assert.doesNotMatch(motifRefus("6600000") ?? "", /1 chiffres/);
  });

  it("nomme les préfixes valides", () => {
    assert.match(motifRefus("51234567") ?? "", /3, 6, 8, 9/);
  });

  it("ne dit rien sur un numéro correct", () => {
    assert.equal(motifRefus("66000000"), null);
    assert.equal(motifRefus("+235 31 23 45 67"), null);
  });

  it("demande une saisie quand le champ est vide", () => {
    assert.notEqual(motifRefus(""), null);
  });
});

describe("Mise en forme", () => {
  it("groupe par deux", () => {
    assert.equal(formatLisible("66000000"), "66 00 00 00");
    assert.equal(formatLisible("+23591912191"), "91 91 21 91");
  });

  it("supporte un numéro incomplet", () => {
    assert.equal(formatLisible("660"), "66 0");
  });
});

describe("Numéros étrangers", () => {
  // Un tuteur peut vivre hors du Tchad. Forcer +235 devant son numéro
  // produirait un destinataire inexistant, et l'alerte d'absence n'arriverait
  // jamais — sans que rien ne le signale.
  it("respecte un indicatif étranger explicite", () => {
    assert.equal(formatInternational("+33612345678"), "+33612345678");
    assert.equal(formatInternational("+225 07 12 34 56 78"), "+2250712345678");
    assert.equal(formatInternational("0033612345678"), "+33612345678");
  });

  it("ramène toujours un numéro tchadien à +235", () => {
    assert.equal(formatInternational("+23566000000"), "+23566000000");
    assert.equal(formatInternational("66000000"), "+23566000000");
    assert.equal(formatInternational("0023566000000"), "+23566000000");
  });

  it("ne considère pas un numéro étranger comme valide ici", () => {
    assert.equal(telephoneValide("+33612345678"), false);
  });
});
